Block supports: Restore root DOMDocument save

sirreal opened the pull request:
Fix a regression from WordPress/gutenberg#25020 which was found to remove anything outside of the block root that may have been appended or prepended. This fix is achieved by concatenating all the childNodes in the DOMDocument body. See https://github.com/WordPress/gutenberg/pull/25026 for additional details on the regression.

Add a test to ensure that appended HTML is preserved.

The regression from #25020 exposes an issue I'll raise in another PR with some failing tests. HTML _appended_ to the block content works well. However, if additional output comes before the block content then then `gutenberg_apply_block_supports` will misbehave because it expects the block content to be the first element in the constructed DOM body:
https://github.com/WordPress/gutenberg/blob/7ee00ad4c5dc64a2b971cebe1d6ee4789eff7157/lib/block-supports/index.php#L75

See https://github.com/WordPress/gutenberg/pull/25053 for discussion around the expected behavior of `gutenberg_apply_block_supports`.

Fixes #25117

sirreal committed:
Revert "Save DOMDocument node instead of root + str_replace (#25020)"

This reverts commit 2473c1d577410bc0af964c1d1cde49c2ae1dff7a.

ockham commented on lib/block-supports/index.php:
```diff
@@ -96,7 +96,7 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 	}
 
-	return $dom->saveHtml( $block_root );
+	return str_replace( array( $wrapper_left, $wrapper_right ), '', $dom->saveHtml() );
```
(That would allow us to retain the functionality we sought to add, and the unit test change.)

sirreal committed:
Trim whitespace from saved DOMDocument HTML

sirreal committed:
Restore surrounding whitespace test

sirreal committed:
Add appended HTML test

sirreal commented:
This is no longer a simple revert. The regression should be addressed by this PR. The whitespace cleanup should be maintained here, and an additional test to cover the regression is introduced.

westonruter commented on lib/block-supports/index.php:
```diff
@@ -96,7 +96,7 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 	}
 
-	return $dom->saveHtml( $block_root );
+	return trim( str_replace( array( $wrapper_left, $wrapper_right ), '', $dom->saveHtml() ) );
```
It seems somewhat unsafe to rely on the original HTML input to be identical after DOM serialization. I'd expect there to be some changes being made like addition of whitespace or changing attribute order or something at some point.

westonruter replied:
You might want to consider what we are doing in the AMP plugin:

https://github.com/ampproject/amp-wp/blob/fc0373defb7cc8e5c4331d57d9140e8bd1454ac3/includes/utils/class-amp-dom-utils.php#L178-L193

sirreal replied:
That's a good point, thanks for bringing it up. I've implemented an alternative that concatenates the child nodes of the body element that I believe improves on the original implementation from https://github.com/WordPress/gutenberg/pull/24645.

sirreal committed:
Concatenate body childNodes

It will be fragile to match exact strings to be removed from a saved
DOMDocument. It's already been observed that this introduces extraneous
newlines and other document normalization could change the strings we
expect to match.

Instead of matching some HTML strings to remove, concatenate the
children of the document body element in order to produce the expected
HTML.

Noted in https://github.com/WordPress/gutenberg/pull/25028#discussion_r483843280

sirreal committed:
Access body element directly via id attribute

sirreal commented:
Addressed comments and rebased to fix conflicts.

sirreal commented:
I'll investigate the test failures…

ockham commented on lib/block-supports/index.php:
```diff
@@ -71,8 +77,8 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
 
-	$xpath      = new DOMXPath( $dom );
-	$block_root = $xpath->query( '/html/body/*' )[0];
+	$body_element = $dom->getElementByID( $body_id );
```
I think it'd be sufficient to 

```suggestion
	$body_element = $dom->getElementsByTagName( 'body' )[0];
```

(which means requiring that no other `<body />` be prepended, which I think is a fair assumption).

Also fine if you wanna play it even safer though :+1:

sirreal replied:
I don't like going through the NodeList to get the first element, but I don't feel strongly.

I've implemented an alternative in b252259e2c52ecff711f812f48d026db700f85b0 that traverses simple, known path to the body.

ockham reviewed (approved):
Looking good, thanks. One non-blocking note [here](https://github.com/WordPress/gutenberg/pull/25028#discussion_r484379944).

Feel free to merge once green!

sirreal committed:
Use `->documentElement->lastChild` to get body

sirreal committed:
Ensure block_root is a DOMElement

sirreal committed:
Use DOMXpath to query children elements

ntsekouras commented:
This will also fix https://github.com/WordPress/gutenberg/issues/25117

sirreal committed:
Add text-only render test

sirreal commented on lib/block-supports/index.php:
```diff
@@ -71,9 +73,13 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
+
 	$xpath      = new DOMXPath( $dom );
-	$block_root = $xpath->query( '/html/body/*' )[0];
+	$block_root = $xpath->query( './*', $body_element )[0];
```
I had incorrectly assumed the xpath `*` query was synonymous to `DOMNode::childNodes`", but it isn't. That was causing the test failures.

The xpath `*` ["selects all element children of the context node"](https://www.w3.org/TR/1999/REC-xpath-19991116/#path-abbrev), whereas `childNodes` may be `DOMNodes` of any type, e.g. text nodes, which later error under some conditions.

I've restored the `DOMXpath` implementation to search for `*` (child elements) in the context of the `$body_element`.

An alternative I considered was to `array_search` with a function to match the `XML_ELEMENT_NODE` DOMNode::nodeType. However, I believe that would require introducing another named function which seems undesirable.

I've introduced a unit test for the text-only render output case.

westonruter commented on lib/block-supports/index.php:
```diff
@@ -98,7 +104,12 @@ function gutenberg_apply_block_supports( $block_content, $block ) {
+	$result = '';
+	// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+	foreach ( $body_element->childNodes as $child_node ) {
+		$result .= $dom->saveHtml( $child_node );
```
The AMP plugin had difficulty using saveHTML in PHP versions prior to 7.3, so have an alternative for older versions:

https://github.com/ampproject/amp-wp/blob/f0b7ca9c9e2aa689be23ad6286cff5a98797f967/lib/common/src/Dom/Document.php#L509-L513

sirreal replied:
I looked into the history around that change a little bit. It looks like it was related to some inconsistent whitespace formatting in the produced HTML. That doesn't seem problematic if its only formatting whitespace, is there anything in particular you think we should watch out for?

westonruter replied:
Yes, it was inconsistent whitespace formatting. But it can indeed be problematic. For example, sometimes you may have markup like:

```html
<ul><li>left</li><li>right</li></ul>
```

If the list items here are displayed `inline`, then this will render differently than:

```html
<ul>
    <li>left</li>
    <li>right</li>
</ul>
```

Namely, there will be a space (" ") between the list items.

sirreal replied:
That does sound like something that will need to be addressed especially as this block-supports work moves towards core. Thanks for flagging.

sirreal replied:
@westonruter I've dug into this a little bit and have been unable to reproduce or to find bug reports. Do you have anything we can go on?

Here's an example:

```php
<?php
// save as test-dom-doc.php

$dom = new DOMDocument();

$html = '<html><head></head><body>
	a text node
<p>a paragraph</p>
<ul>
<li>a</li>
	<li>list</li>
		<li>is</li>
			<li>ordered
<em>emphasis</em></li>

</ul>
<pre>       obviously
 this     whitespace

is
    important</pre>
another     text    node with a bunch of wierd space
</body>
</html>';

$dom->loadHtml( $html );

$body = $dom->documentElement->lastChild;

$result = '';
foreach ( $body->childNodes as $c ) {
	$result .= $dom->saveHtml( $c );
}
echo $result;
```

```sh
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:5-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:7.0-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:7.4-cli php test-dom-doc.php
docker run -it --rm --name check-domdoc-ws -v "$PWD":/usr/src/myapp -w /usr/src/myapp php:8.0.0beta3-cli php test-dom-doc.php
```

All of these seem to produce identical output with whitespace:

```html

	a text node
<p>a paragraph</p>
<ul><li>a</li>
	<li>list</li>
		<li>is</li>
			<li>ordered
<em>emphasis</em></li>

</ul><pre>       obviously
 this     whitespace

is
    important</pre>
another     text    node with a bunch of wierd space
```

[This bug report](https://bugs.php.net/bug.php?id=78129) looks similar, but has apparently stalled and [was suggested](https://bugs.php.net/bug.php?id=78129#1591607348) it was caused by an outdated libxml2.

westonruter replied:
Here you go: https://3v4l.org/Adldb

sirreal replied:
That's perfect, thank you.

sirreal replied:
I created a PR to address this: https://github.com/WordPress/gutenberg/pull/25240
